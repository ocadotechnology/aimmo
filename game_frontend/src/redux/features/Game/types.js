const SOCKET_CONNECT_TO_GAME_REQUEST = 'features/Game/SOCKET_CONNECT_TO_GAME_REQUEST'
const SOCKET_CONNECT_TO_GAME_FAIL = 'features/Game/SOCKET_CONNECT_TO_GAME_FAIL'

const SOCKET_GAME_STATE_RECEIVED = 'features/Game/SOCKET_GAME_STATE_RECEIVED'
const SEND_GAME_STATE_SUCCESS = 'features/Game/SOCKET_GAME_STATE_UPDATE_SUCCESS'
const SEND_GAME_STATE_FAIL = 'features/Game/SOCKET_GAME_STATE_UPDATE_FAIL'

const UNITY_EVENT = 'features/Game/UNITY_EVENT'

const CONNECTION_PARAMETERS_RECEIVED = 'features/Game/CONNECTION_PARAMETERS_RECEIVED'

const UNITY_SEND_AVATAR_ID_SUCCESS = 'features/Game/UNITY_SEND_AVATAR_ID_SUCCESS'
const UNITY_SEND_AVATAR_ID_FAIL = 'features/Game/UNITY_SEND_AVATAR_ID_FAIL'

const SOCKET_FEEDBACK_AVATAR_UPDATED = 'features/Game/SOCKET_FEEDBACK_AVATAR_UPDATED'
const SNACKBAR_FOR_AVATAR_FEEDBACK_SHOWN = 'features/Game/SNACKBAR_FOR_AVATAR_FEEDBACK_SHOWN'

const GAME_DATA_LOADED = 'features/Game/GAME_DATA_LOADED'

const SET_TIMEOUT = 'features/Game/TIMEOUT'

export default {
  SOCKET_CONNECT_TO_GAME_REQUEST,
  SOCKET_CONNECT_TO_GAME_FAIL,
  SOCKET_GAME_STATE_RECEIVED,
  SEND_GAME_STATE_SUCCESS,
  SEND_GAME_STATE_FAIL,
  UNITY_EVENT,
  CONNECTION_PARAMETERS_RECEIVED,
  UNITY_SEND_AVATAR_ID_SUCCESS,
  UNITY_SEND_AVATAR_ID_FAIL,
  SOCKET_FEEDBACK_AVATAR_UPDATED,
  SNACKBAR_FOR_AVATAR_FEEDBACK_SHOWN,
  GAME_DATA_LOADED,
  SET_TIMEOUT
}
