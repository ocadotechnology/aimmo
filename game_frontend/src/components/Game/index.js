import React, { Component, Fragment } from 'react'
import GameMenu from 'components/GameMenu'
import GameView from 'components/GameView'

export default class Game extends Component {
  render () {
    return (
      <Fragment>
        <GameMenu />
        <GameView />
      </Fragment>
    )
  }
}
