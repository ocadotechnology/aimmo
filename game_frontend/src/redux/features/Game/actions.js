import types from './types'

const getConnectionParametersRequest = gameID => (
  {
    type: types.GET_CONNECTION_PARAMETERS_REQUEST,
    payload: {
      gameID
    }
  }
)

const getConnectionParametersSuccess = connectionParameters => (
  {
    type: types.GET_CONNECTION_PARAMETERS_SUCCESS,
    payload: {
      connectionParameters
    }
  }
)

const setGameURL = gameURL => (
  {
    type: types.SET_GAME_URL,
    payload: {
      gameURL
    }
  }
)

const setGameURLSuccess = () => (
  {
    type: types.SET_GAME_URL_SUCCESS
  }
)

const setGameURLFail = error => (
  {
    type: types.SET_GAME_URL_FAIL,
    payload: {
      error
    }
  }
)

const setGamePath = gamePath => (
  {
    type: types.SET_GAME_PATH,
    payload: {
      gamePath
    }
  }
)

const setGamePathSuccess = () => (
  {
    type: types.SET_GAME_PATH_SUCCESS
  }
)

const setGamePathFail = error => (
  {
    type: types.SET_GAME_PATH_FAIL,
    payload: {
      error
    }
  }
)

const setGamePort = gamePort => (
  {
    type: types.SET_GAME_PORT,
    payload: {
      gamePort
    }
  }
)

const setGamePortSuccess = () => (
  {
    type: types.SET_GAME_PORT_SUCCESS
  }
)

const setGamePortFail = error => (
  {
    type: types.SET_GAME_PORT_FAIL,
    payload: {
      error
    }
  }
)

const setGameSSL = gameSSLFlag => (
  {
    type: types.SET_GAME_SSL,
    payload: {
      gameSSLFlag
    }
  }
)

const setGameSSLSuccess = () => (
  {
    type: types.SET_GAME_SSL_SUCCESS
  }
)

const setGameSSLFail = error => (
  {
    type: types.SET_GAME_SSL_FAIL,
    payload: {
      error
    }
  }
)

const establishGameConnection = () => (
  {
    type: types.ESTABLISH_GAME_CONNECTION
  }
)

const establishGameConnectionSuccess = () => (
  {
    type: types.ESTABLISH_GAME_CONNECTION_SUCCESS
  }
)

const establishGameConnectionFail = error => (
  {
    type: types.ESTABLISH_GAME_CONNECTION_FAIL,
    payload: {
      error
    }
  }
)

const unityEvent = (unityEvent, unityData, successAction, failAction) => (
  {
    type: types.UNITY_EVENT,
    payload: {
      unityEvent,
      unityData,
      successAction,
      failAction
    }
  }
)

export default {
  getConnectionParametersRequest,
  getConnectionParametersSuccess,
  setGameURL,
  setGameURLSuccess,
  setGameURLFail,
  setGamePath,
  setGamePathSuccess,
  setGamePathFail,
  setGamePort,
  setGamePortSuccess,
  setGamePortFail,
  setGameSSL,
  setGameSSLSuccess,
  setGameSSLFail,
  establishGameConnection,
  establishGameConnectionSuccess,
  establishGameConnectionFail,
  unityEvent
}
