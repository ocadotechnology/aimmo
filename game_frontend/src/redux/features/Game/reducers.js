import types from './types'

const gameReducer = (state = {}, action) => {
  switch (action.type) {
    case types.GET_CONNECTION_PARAMETERS_SUCCESS:
      return {
        ...state,
        connectionParameters: action.payload.connectionParameters
      }
    default:
      return state
  }
}

export default gameReducer
