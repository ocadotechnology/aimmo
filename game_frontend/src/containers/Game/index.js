import React, { Component, Fragment } from 'react'
import GameView from 'components/GameView'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { actions } from 'redux/features/Game'
import Snackbar from 'components/Snackbar'

export class Game extends Component {
  static propTypes = {
    connectToGame: PropTypes.func,
    theme: PropTypes.object
  }

  state = {
    showSnackbar: this.props.showSnackbarForAvatarUpdated
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.showSnackbarForAvatarUpdated !== this.props.showSnackbarForAvatarUpdated) {
      this.setState({
        showSnackbar: nextProps.showSnackbarForAvatarUpdated
      })
    }
  }

  handleClose = () => {
    this.setState({ showSnackbar: false })
    this.props.snackbarForAvatarUpdatedShown()
  }

  render () {
    return (
      <Fragment>
        <GameView connectToGame={this.props.connectToGame} />
        <Snackbar
          type='success'
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
          open={this.state.showSnackbar}
          direction='up'
          onClose={this.handleClose}
          message='Your Avatar has been updated with your new code!'
        />
      </Fragment>
    )
  }
}

const mapDispatchToProps = {
  connectToGame: actions.socketConnectToGameRequest,
  snackbarForAvatarUpdatedShown: actions.snackbarForAvatarUpdatedShown
}

const mapStateToProps = state => ({
  showSnackbarForAvatarUpdated: state.game.showSnackbarForAvatarUpdated
})

export default connect(mapStateToProps, mapDispatchToProps)(Game)
