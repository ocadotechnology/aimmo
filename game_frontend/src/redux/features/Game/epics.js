import actions from './actions'
import types from './types'
import { Observable } from 'rxjs'
import { map, mergeMap, catchError, delay, tap } from 'rxjs/operators'
import { ofType } from 'redux-observable'

const getConnectionParametersEpic = (action$, store, { api: { get } }) => action$.pipe(
  ofType(types.SOCKET_CONNECT_TO_GAME_REQUEST),
  mergeMap(action =>
    get(`games/${store.getState().game.connectionParameters.game_id}/connection_parameters/`).pipe(
      map(response => actions.connectionParametersReceived(response))
    )
  )
)

const sendGameStateEpic = (action$, store, { api: { unity } }) => action$.pipe(
  ofType(types.SOCKET_GAME_STATE_RECEIVED),
  map(action => actions.unityEvent(
    'ReceiveGameUpdate',
    JSON.stringify(action.payload.gameState),
    actions.sendGameStateSuccess(),
    actions.sendGameStateFail
  )),
  unity.sendExternalEvent(unity.emitToUnity)
)

const connectToGameEpic = (action$, store, { api: { socket, unity } }) => action$.pipe(
  ofType(types.CONNECTION_PARAMETERS_RECEIVED),
  socket.connectToGame(),
  socket.startListeners(),
  catchError(error => Observable.of({
    type: types.SOCKET_CONNECT_TO_GAME_FAIL,
    payload: error,
    error: true
  }))
)

const sendAvatarIDEpic = (action$, store, { api: { unity } }) => action$.pipe(
  ofType(types.CONNECTION_PARAMETERS_RECEIVED),
  map(action => actions.unityEvent(
    'SetCurrentAvatarID',
    parseInt(action.payload.parameters['avatar_id']),
    actions.unitySendAvatarIDSuccess(),
    actions.unitySendAvatarIDFail
  )),
  unity.sendExternalEvent(unity.emitToUnity)
)

export default {
  getConnectionParametersEpic,
  connectToGameEpic,
  sendGameStateEpic,
  sendAvatarIDEpic
}
