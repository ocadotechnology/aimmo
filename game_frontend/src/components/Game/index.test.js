/* eslint-env jest */
import React from 'react'
import Game from 'components/Game'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('<Game />', () => {
  it('renders correctly', () => {
    const tree = shallow(<Game />)
    expect(toJson(tree)).toMatchSnapshot()
  })
})
