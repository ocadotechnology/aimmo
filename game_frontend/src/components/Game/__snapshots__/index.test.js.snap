// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> renders correctly 1`] = `
<React.Fragment>
  <styled.nav />
  <styled.div />
</React.Fragment>
`;
